### ISSUE TITLE
Exception when killing calculation

### ISSUE THREAD

**Issue by greschd** [aiidateam/aiida-core#3888]
When killing a calculation, I consistently get the following error in the log:

```
04/02/2020 10:05:18 AM <17123> kiwipy.rmq.tasks: [ERROR] The task excepted
Traceback (most recent call last):
  File "/data/aiida/source/aiida-core/aiida/manage/external/rmq.py", line 187, in _continue
    result = yield super()._continue(communicator, pid, nowait, tag)
  File "/home/a-dogres/.virtualenvs/aiida/lib/python3.6/site-packages/tornado/gen.py", line 1055, in run
    value = future.result()
  File "/home/a-dogres/.virtualenvs/aiida/lib/python3.6/site-packages/tornado/concurrent.py", line 238, in result
    raise_exc_info(self._exc_info)
  File "<string>", line 4, in raise_exc_info
  File "/home/a-dogres/.virtualenvs/aiida/lib/python3.6/site-packages/tornado/gen.py", line 1069, in run
    yielded = self.gen.send(value)
  File "/home/a-dogres/.virtualenvs/aiida/lib/python3.6/site-packages/plumpy/process_comms.py", line 554, in _continue
    raise gen.Return(proc.future().result())
  File "/home/a-dogres/.virtualenvs/aiida/lib/python3.6/site-packages/plumpy/futures.py", line 34, in result
    return super(Future, self).result(timeout)
  File "/home/a-dogres/.virtualenvs/aiida/lib/python3.6/site-packages/tornado/concurrent.py", line 238, in result
    raise_exc_info(self._exc_info)
  File "<string>", line 4, in raise_exc_info
plumpy.exceptions.KilledError: Killed through `verdi process kill`

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/a-dogres/.virtualenvs/aiida/lib/python3.6/site-packages/kiwipy/rmq/tasks.py", line 126, in _on_task
    result = yield result
  File "/home/a-dogres/.virtualenvs/aiida/lib/python3.6/site-packages/tornado/gen.py", line 1055, in run
    value = future.result()
  File "/home/a-dogres/.virtualenvs/aiida/lib/python3.6/site-packages/tornado/concurrent.py", line 238, in result
    raise_exc_info(self._exc_info)
  File "<string>", line 4, in raise_exc_info
  File "/home/a-dogres/.virtualenvs/aiida/lib/python3.6/site-packages/kiwipy/futures.py", line 54, in capture_exceptions
    yield
  File "/home/a-dogres/.virtualenvs/aiida/lib/python3.6/site-packages/plumpy/futures.py", line 103, in run_task
    future.set_result((yield coro()))
  File "/home/a-dogres/.virtualenvs/aiida/lib/python3.6/site-packages/tornado/gen.py", line 1055, in run
    value = future.result()
  File "/home/a-dogres/.virtualenvs/aiida/lib/python3.6/site-packages/tornado/concurrent.py", line 238, in result
    raise_exc_info(self._exc_info)
  File "<string>", line 4, in raise_exc_info
  File "/home/a-dogres/.virtualenvs/aiida/lib/python3.6/site-packages/tornado/gen.py", line 1063, in run
    yielded = self.gen.throw(*exc_info)
  File "/home/a-dogres/.virtualenvs/aiida/lib/python3.6/site-packages/plumpy/process_comms.py", line 492, in __call__
    raise gen.Return((yield self._continue(communicator, **task.get(TASK_ARGS, {}))))
  File "/home/a-dogres/.virtualenvs/aiida/lib/python3.6/site-packages/tornado/gen.py", line 1055, in run
    value = future.result()
  File "/home/a-dogres/.virtualenvs/aiida/lib/python3.6/site-packages/tornado/concurrent.py", line 238, in result
    raise_exc_info(self._exc_info)
  File "<string>", line 4, in raise_exc_info
  File "/home/a-dogres/.virtualenvs/aiida/lib/python3.6/site-packages/tornado/gen.py", line 1063, in run
    yielded = self.gen.throw(*exc_info)
  File "/data/aiida/source/aiida-core/aiida/manage/external/rmq.py", line 194, in _continue
    self.handle_continue_exception(node, exception, message)
  File "/data/aiida/source/aiida-core/aiida/manage/external/rmq.py", line 137, in handle_continue_exception
    node.set_exception(str(exception))
  File "/data/aiida/source/aiida-core/aiida/orm/nodes/process/process.py", line 359, in set_exception
    return self.set_attribute(self.EXCEPTION_KEY, exception)
  File "/data/aiida/source/aiida-core/aiida/orm/utils/mixins.py", line 176, in set_attribute
    raise exceptions.ModificationNotAllowed('attributes of a sealed node are immutable')
aiida.common.exceptions.ModificationNotAllowed: attributes of a sealed node are immutable
```

Curiously, the process _is_ in `Killed` state after all that. I suspected multiple daemon workers trying to put it into `Killed` state, but this happens also with a single worker. I can also reproduce this both locally and on the production instance.

Not sure if this is relevant, but (all?) log messages also show up duplicated.

@sphuber do you have an idea for how to further debug this?

**Comment by sphuber** [comment]
> Not sure if this is relevant, but (all?) log messages also show up duplicated.

That I also noticed a while back, but should not be related. I forgot to open an issue, but just did see #3889.

Not sure yet how to debug this further, but this might be related to issue #2861 . Currently if an exception occurs in on of the state transition handlers, e.g. `on_entered` it will try to go to `Excepted` but if it already was put in a terminal state, then this will fail. This will need to be fixed in `plumpy`

**Comment by greschd** [comment]
That would still leave figuring out why an exception occurs in the state transition handler in the first place, right?

I'll do some more prodding around..

**Comment by sphuber** [comment]
> That would still leave figuring out why an exception occurs in the state transition handler in the first place, right?
> 
> I'll do some more prodding around..

yes it would :smile:

**Comment by oashour** [comment]
I am suffering from this issue at the moment and my daemon refuses to start backup. Everything I try to do --start, stop or reset-- just times out. Rebooting and deactivating/reactivating the venv did not work. Any thoughts?

**Comment by greschd** [comment]
That _sounds_ unrelated from the face of it - I didn't have any issues with an unstartable daemon.

Could you open a separate issue and provide some more information?

**Comment by chrisjsewell** [comment]
I believe I may have identified the cause of this bug in https://github.com/aiidateam/plumpy/pull/180
See https://github.com/aiidateam/plumpy/blob/942f22c6ff9526b8718f3c46fcb6d1dd8aa8a5f3/plumpy/process_states.py#L253-L254

**Comment by unkcpz** [comment]
@chrisjsewell Yes, I dos't think it is related but it reminds me that the second part of this issue, the node seal exception will be fixed by the following change, adding a `node.is_sealed` in the exception handling. If so, will also make a PR for the current tornado version `aiida_core`

https://github.com/unkcpz/aiida_core/blob/8cfadb083ddb41df443ca1618670de0ab4bc87e9/aiida/manage/external/rmq.py#L155

**Comment by unkcpz** [comment]
As for the issue itself, if I understand it correctly, the exception is correctly raised here and it is propagated from the  `future` capture and raised by `kiwipy.capture_exceptions(future)` in `plumpy/future:create_task()`. All we need to do(not sure and don't know how to do it) is to elegantly handle this exception and print a clean message in the logs.

**Comment by chrisjsewell** [comment]
so this is the cause (or at least one):

1. in `kill` the parent is killed first, then the children: https://github.com/aiidateam/aiida-core/blob/8c079175bbb32a6dc3b4ae663c586c7099cbb9a6/aiida/engine/processes/process.py#L294
2. when entering the `Wait` state awaitables (e.g. children) are each assigned to `on_process_finished`: https://github.com/aiidateam/aiida-core/blob/8c079175bbb32a6dc3b4ae663c586c7099cbb9a6/aiida/engine/processes/workchains/workchain.py#L263
3. this calls `resolve_awaitable`, which tries to update the status: https://github.com/aiidateam/aiida-core/blob/8c079175bbb32a6dc3b4ae663c586c7099cbb9a6/aiida/engine/processes/workchains/workchain.py#L145

so basically in `resolve_awaitable` we want to check if the parent is already in a terminal state and, if so, skip the status update and probably just everything after `self._awaitables.remove(awaitable)` (still good to do this, to remove cyclic references)

**Comment by chrisjsewell** [comment]
#4773 is now merged and I've opened #4775 to discuss about such exceptions can be better identified during testing.

@greschd and @unkcpz, can you let me know if this fixes the issue and if you think this issue can now be closed?

**Comment by greschd** [comment]
I can't really test that on the production instance where I get these errors. If you've tried killing a workchain after the fix and didn't get an exception I'd say we close this, and re-open if there's another case.

**Comment by chrisjsewell** [comment]
sounds good to me
